feat: extension can register custom kube generator

axel7083 opened the pull request:
### What does this PR do?

The default way of generating kubernetes files from a container/group of containers is not use the `generatePodmanKube`.

This PR aims to add the possibility for extensions to add their own `KubeGenerator`, allowing to register a command, that will be called when we want to generate a Kubernetes Config.
### Example

See bellow in comment 

### Screenshot/screencast of this PR

![image](https://github.com/containers/podman-desktop/assets/42176370/ce482c1e-a9c9-4f19-984f-6ce205401e13)

### What issues does this PR fix or reference?

<!-- Include any related issues from Podman Desktop 
repository (or from another issue tracker). -->

### How to test this PR?

<!-- Please explain steps to reproduce -->

benoitf commented:
hello, I think here we should use the code to declare generators rather than declare contributes

like `registerKubernetesGeneratorProvider` (or something with the same meaning) in the API

then we could move the embedded Generator in a separate extension and then you could disable the 'default one', provide new ones, and if there are multiple providers, you would be prompted to select the one you want

slemeur commented:
Super cool contribution @axel7083 !

axel7083 commented:
> hello, I think here we should use the code to declare generators rather than declare contributes
> 
> like `registerKubernetesGeneratorProvider` (or something with the same meaning) in the API
> 
> then we could move the embedded Generator in a separate extension and then you could disable the 'default one', provide new ones, and if there are multiple providers, you would be prompted to select the one you want

I removed the code responsible of loading the generator from the `package.json` and replaced it with the code declarative way.

E.g
```ts
import * as podmanDesktopAPI from '@podman-desktop/api';

export async function activate(extensionContext: podmanDesktopAPI.ExtensionContext): Promise<void> {
  extensionContext.subscriptions.push(
      podmanDesktopAPI.kubernetesGenerator.registerKubernetesGeneratorProvider({
          command: 'kompose.generateKube',
          id: 'kompose',
      }),
      podmanDesktopAPI.commands.registerCommand(
          'kompose.generateKube',
          async (engineId: string, names: string[]): Promise<string> => {
              console.log("kompose.generateKube", engineId, names);
              return "hello from extension";
          }),
  );
}
```

benoitf commented:
you shouldn't need to register a command, it's the provider that will handle the generation part
probably it can be in the `kubernetes` scope 

`podmanDesktopAPI.kubernetes.registerKubernetesGeneratorProvider`

and then you don't give a command but the callback itself

registerKubernetesGeneratorProvider(provider)

and provider has API like

```
{
   generate(engineId: string, ids: string[]): GenerateKubeResult
}
```

and in the result you have for example a `.yaml` field

axel7083 commented:
> you shouldn't need to register a command, it's the provider that will handle the generation part probably it can be in the `kubernetes` scope
> 
> `podmanDesktopAPI.kubernetes.registerKubernetesGeneratorProvider`
> 
> and then you don't give a command but the callback itself
> 
> registerKubernetesGeneratorProvider(provider)
> 
> and provider has API like
> 
> ```
> {
>    generate(engineId: string, ids: string[]): GenerateKubeResult
> }
> ```
> 
> and in the result you have for example a `.yaml` field

Okey, I made the changes. Here is the current way for registering a custom KubeGenerator
```ts
import * as podmanDesktopAPI from '@podman-desktop/api';

export async function activate(extensionContext: podmanDesktopAPI.ExtensionContext): Promise<void> {
  extensionContext.subscriptions.push(
      podmanDesktopAPI.kubernetes.registerKubernetesGeneratorProvider({
          id: 'kompose',
          generate: (engineId: string, ids: string[]) => {
              return {
                  yaml: "hello from extension"
              };
          }
      }),
  );
}
```

### Next step

- [ ] Renaming `KubeGenerator` to `KubernetesGenerator`
- [ ] Creating a unique Svelte component instead of having `ContainerDetailsKube.svelte`, `PodDetailsKube.svelte`, `ComposeDetailsKube.svelte`
- [ ] Exploring solution to select the provider to display in Kube
- [ ] Creating unit tests

@benoitf tell me if I should take some part in another PR

benoitf commented on packages/extension-api/src/extension-api.d.ts:
```diff
@@ -1330,6 +1330,18 @@ declare module '@podman-desktop/api' {
+     * Add a KubernetesGenerator to KubernetesGeneratorRegistry
+     * @param kubeGenerator the generator to add
+     */
+    export function registerKubernetesGeneratorProvider(kubeGenerator: KubeGenerator): Disposable;
```
where is declared `KubeGenerator` ?

benoitf replied:
hint: all resources involved need to be in the .d.ts file

benoitf commented on packages/extension-api/src/extension-api.d.ts:
```diff
@@ -1330,6 +1330,18 @@ declare module '@podman-desktop/api' {
+     * @param kubeGenerator the generator to add
+     */
+    export function registerKubernetesGeneratorProvider(kubeGenerator: KubeGenerator): Disposable;
+
+    /**
+     * Remove a provider from the KubernetesGeneratorRegistry
+     * @param kubeGenerator the generator to remove
+     */
+    export function unregisterKubernetesGeneratorProvider(kubeGenerator: KubeGenerator): void;
```
you can delete the unregister

as register returns a disposable resource, when calling `dispose()` it'll unregister the provider

benoitf commented on packages/main/src/plugin/kube-generator-registry.ts:
```diff
@@ -0,0 +1,48 @@
+export interface GenerateKubeResult {
+  yaml: string;
+}
+export interface Provider {
```
you need to add a prefix as Provider is too generic

benoitf commented:
> Next step
 Renaming KubeGenerator to KubernetesGenerator
 Creating a unique Svelte component instead of having ContainerDetailsKube.svelte, PodDetailsKube.svelte, ComposeDetailsKube.svelte
 Exploring solution to select the provider to display in Kube
 Creating unit tests


I would do renaming in a separate PR (first one)

Second one would be
 Creating a unique Svelte component instead of having ContainerDetailsKube.svelte, PodDetailsKube.svelte, ComposeDetailsKube.svelte
+ tests

Another PR is ok with the registration of the providers (even if not used)
I think it's missing the stores object where client side can know the list of kube generator providers being registered
so api.d.ts, registration, store
and unit tests for that
(third PR)

then, update UI to use these providers/ select deselect

And probably I would add another PR that would be to move out to a separate extension the podman-kube provider

(so you can disable this generator provider if you don't want this one)


and having all PRs with their tests

cdrage commented:
As someone who's worked on Kompose for a long time, this makes me happy @axel7083 and thanks for the contribution!

I agree with all of @benoitf points with regards to extension updates / unit tests / exploring solution to select a provider. 

Having this added as part of the kubernetes generation has been on our roadmap and this looks like an elegant solution to generate the Kubernetes YAML from a group of containers.

The biggest problem I see right now is that **this will not work for single containers or pods** and will only work for compsoe files.

Since kompose takes in only docker-compose YAML...

This is something I brought up in this issue https://github.com/kubernetes/kompose/issues/1570

benoitf commented:
so somehow we need to add tweaks in the provider, etc to say that this kind of generator will only apply for a given set of resources

axel7083 commented:
> so somehow we need to add tweaks in the provider, etc to say that this kind of generator will only apply for a given set of resources

Adding when property ? Something like some conditions on a context specific to container type ?

benoitf commented:
I think the Provider object can have a method `support` or `accept` where we give a enum like 'Container', 'Pod', 'Compose' and it returns a boolean.

Then, when we ask for a provider, we'll iterate on all generators supporting the given type

for example if we have containers but Kompose generator is not handling it, `accept(acceptContext)` would return false
in acceptContext we have like `type: Container` and it allows to grow the acceptance criteria

in your current PR, d.ts file is still missing the interface of a Generator

and I think this one is considered as the 3rd one so we probably need to have/merge 1. and 2. before

benoitf commented:
there could be also a selector when registering the provider that will tell on which resources it applies
so it's when we register the resource, not on each call (faster at runtime)

axel7083 commented:
> I think the Provider object can have a method `support` or `accept` where we give a enum like 'Container', 'Pod', 'Compose' and it returns a boolean.
> 
> Then, when we ask for a provider, we'll iterate on all generators supporting the given type
> 
> for example if we have containers but Kompose generator is not handling it, `accept(acceptContext)` would return false in acceptContext we have like `type: Container` and it allows to grow the acceptance criteria
> 
> in your current PR, d.ts file is still missing the interface of a Generator
> 
> and I think this one is considered as the 3rd one so we probably need to have/merge 1. and 2. before

I implemented a very basic `accept?(type: KubernetesGeneratorType): boolean;` for now with `KubernetesGeneratorType` being an enum `COMPOSE | POD | Container`, since I was not sure, what you want the acceptContext object to be ? Do we use `ContextUI` ?

benoitf commented:
baiscally you do
```ts
registerKubernetesGenerator(selector: KubernetesGeneratorSelector, provider: KubeGenProvider);
```

and KubernetesGeneratorSelector is

```ts
 KubernetesGeneratorType | ReadonlyArray<KubernetesGeneratorType[]>
```


so Kompose extension will do

```ts
api.kubernetes.registerKubernetesGenerator('Compose', provider);
```

axel7083 commented:
> baiscally you do
> 
> ```ts
> registerKubernetesGenerator(selector: KubernetesGeneratorSelector, provider: KubeGenProvider);
> ```
> 
> and KubernetesGeneratorSelector is
> 
> ```ts
>  KubernetesGeneratorType | ReadonlyArray<KubernetesGeneratorType[]>
> ```
> 
> so Kompose extension will do
> 
> ```ts
> api.kubernetes.registerKubernetesGenerator('Compose', provider);
> ```

Okey, I adapted the code to do that.

I was using an enum at first for `KubernetesGeneratorType` and I changed to a type to avoid very long and ugly imports in svelte components.

- [x] The `window.getKubeGeneratorsIdsByType` will return a list of ids (string[]) for a give type. (E.g `await window.getKubeGeneratorsIdsByType('Compose')`
- [x] The `KubeGeneratorRegistry` has a new data structure to store the selector associated with the Providers. (The `registerKubeGenerator` and `unregisterKubeGenerator` methods have been adapted).

benoitf commented on packages/main/src/plugin/index.ts:
```diff
@@ -825,8 +828,18 @@ export class PluginSystem {
+        names: string[],
+        kubeGeneratorId: string | undefined = undefined,
+      ): Promise<string> => {
+        if (!kubeGeneratorId) return containerProviderRegistry.generatePodmanKube(engine, names);
+
+        const kubeGenerator = kubeGeneratorRegistry.getKubeGenerator(kubeGeneratorId);
+        if (!kubeGenerator) throw new Error(`kubeGenerator with id ${kubeGenerator} cannot be found.`);
+
+        return kubeGenerator.generate(engine, names).yaml;
```
I think it would be an area of improvement as I guess that the generatorId might reference a generator that is not supported for the call

So I think we should add the 'type[]' parameter 

Like if I want to generate for a Pod, a Container, a Compose, or both Pods and Containers, etc.

axel7083 replied:
Who is responsible of checking if the generator is compatible with the type ? The client (E.g. svelte component) or the regitry ?

benoitf replied:
client will do a filter to have a better UX

like: you probably don't want to display a generator that will not work later when you click on it saying 'not supported'

but you check on the 'server/main' side as well as you should never trust clients ;-)


so client side = checks for UX experience
server side = checks to not let failures happen

axel7083 replied:
How do I ensure the argument `kubeGeneratorId` when the client call `generatePodmanKube` is calling for the right type ? If the type is provided by the client, nothing prevent the client from sending a kubeGeneratorId and a type compatible, but for the wrong images ids.

benoitf replied:
for the right type, either we remove it by computing it from the names provided

(for example if id is a container's id it will be a 'Container' type being asked) and if it's a Pod id


but maybe we should change the signature to match resources like

we don't ask for names (that are ids like pod's id, container's id, etc) but client provide the resource id like

```
{
containers: [1234, 456]
pods: [123, 345]
compose: [a1, a2]
}
```

so on the main side we're able to verify that pods being asked 123 exists and if it doesn't we abort

benoitf replied:
we provide a more structured object rather than names and we provide the id that should be used

so we can validate if all ids are matching what they are and validate if the provided id would be able to generate kube for the wanted resources

benoitf commented on packages/main/src/plugin/index.ts:
```diff
@@ -1131,6 +1144,13 @@ export class PluginSystem {
       return menuRegistry.getContributedMenus(context);
     });
 
+    this.ipcHandle(
+      'kube-generator-registry:getKubeGeneratorsIdsByType',
+      async (_, type: KubernetesGeneratorType): Promise<string[]> => {
+        return kubeGeneratorRegistry.getKubeGeneratorsIdsByType(type);
+      },
+    );
+
```
it should return an array of KubeGeneratorsInfo objects

in the Info, it should contains the id/name but also the supported types.

axel7083 replied:
Why adding a name property ? For the UI ? Can't we just use the id for that ?

benoitf replied:
so usually id is generated like a counter and name is provided by the provider

from the UI, you want to display `Do you want to generate using the Kompose Generator` ? and then you'll submit the id (like 0, 1 or uuid) to identify the resource

if we have only ids you'll display `Do you want to generate using the 12aa16 Generator ?`

axel7083 replied:
Why would we have the id generated here ? You want to remove the possibility for an extension to choose the id and provide only the name ?

benoitf replied:
an example here
https://github.com/containers/podman-desktop/blob/main/packages/main/src/plugin/provider-registry.ts#L173-L182

(dumb approach)

benoitf commented on packages/main/src/plugin/kube-generator-registry.ts:
```diff
@@ -0,0 +1,67 @@
+export interface KubernetesGeneratorProvider {
+  id: string;
+  generate(engineId: string, ids: string[]): GenerateKubeResult;
+}
+
+export class KubeGeneratorRegistry {
+  private kubeGenerators = new Map<string, KubernetesGeneratorProvider>();
+  private kubeGeneratorsSelectors = new Map<KubernetesGeneratorType, string[]>();
+
+  unregisterKubeGenerator(kubeGenerator: KubernetesGeneratorProvider): void {
+    this.kubeGenerators.delete(kubeGenerator.id);
+
+    for (const [key, value] of this.kubeGeneratorsSelectors) {
+      const updatedIds = value.filter(existingId => existingId !== kubeGenerator.id);
+      this.kubeGeneratorsSelectors.set(key, updatedIds);
+    }
+  }
+
+  registerKubeGenerator(selector: KubernetesGeneratorSelector, kubeGenerator: KubernetesGeneratorProvider): Disposable {
+    this.kubeGenerators.set(kubeGenerator.id, kubeGenerator);
+
+    const selectors = Array.isArray(selector) ? selector : [selector];
+
+    for (const s of selectors) {
+      this.kubeGeneratorsSelectors.set(s, [...(this.kubeGeneratorsSelectors.get(s) ?? []), kubeGenerator.id]);
+    }
+
+    return Disposable.create(() => {
+      this.unregisterKubeGenerator(kubeGenerator);
+    });
+  }
+
+  getKubeGenerator(kubeGeneratorId: string): KubernetesGeneratorProvider | undefined {
+    return this.kubeGenerators.get(kubeGeneratorId);
+  }
+
+  getKubeGeneratorsIdsByType(type: KubernetesGeneratorType): string[] {
+    return this.kubeGeneratorsSelectors.get(type) ?? [];
+  }
+}
```
it requires unit tests

benoitf commented on packages/preload/src/index.ts:
```diff
@@ -923,6 +927,13 @@ function initExposure(): void {
     return ipcInvoke('menu-registry:getContributedMenus', context);
   });
 
+  contextBridge.exposeInMainWorld(
+    'getKubeGeneratorsIdsByType',
+    async (type: KubernetesGeneratorType): Promise<string[]> => {
+      return ipcInvoke('kube-generator-registry:getKubeGeneratorsIdsByType', type);
+    },
+  );
```
we need a more rich structure on the client side, not only the id but also the supported type.

So for example we know if we have multiple generators supporting Pods or none, etc.

axel7083 replied:
Does replacing `KubernetesGeneratorType` with something like `KubernetesGenerators` wthout arguments, and simply returning an array of KubeGeneratorsInfo (name, id, types) is better ?

It would let the client filtering on their need.

benoitf replied:
so expose `getKubeGenerators()` returning `KubeGeneratorsInfo[]` object having `name, id and types` property.

but it's fine to keep types[] as a parameter like client could request only Generators matching a given subset of types (optional arguments)
But that's fine to not provide the argument as well

benoitf commented on packages/preload/src/index.ts:
```diff
@@ -245,9 +246,12 @@ function initExposure(): void {
   });
-  contextBridge.exposeInMainWorld('generatePodmanKube', async (engine: string, names: string[]): Promise<string> => {
-    return ipcInvoke('container-provider-registry:generatePodmanKube', engine, names);
-  });
+  contextBridge.exposeInMainWorld(
+    'generatePodmanKube',
+    async (engine: string, names: string[], kubeGeneratorId: string | undefined = undefined): Promise<string> => {
+      return ipcInvoke('container-provider-registry:generatePodmanKube', engine, names, kubeGeneratorId);
+    },
+  );
```
rename will occur in another PR, right ?

benoitf commented on packages/main/src/plugin/index.ts:
```diff
@@ -825,8 +829,18 @@ export class PluginSystem {
+        _listener,
+        engine: string,
+        names: string[],
+        kubeGeneratorId: string | undefined = undefined,
```
```suggestion
        kubeGeneratorId?: string,
```

benoitf commented on packages/main/src/plugin/index.ts:
```diff
@@ -1131,6 +1145,13 @@ export class PluginSystem {
 
+    this.ipcHandle(
+      'kube-generator-registry:getKubeGeneratorsInfos',
+      async (_, selector: KubernetesGeneratorSelector | undefined = undefined): Promise<KubeGeneratorsInfo[]> => {
```
```suggestion
      async (_, selector?: KubernetesGeneratorSelector): Promise<KubeGeneratorsInfo[]> => {
```

benoitf commented on packages/main/src/plugin/kube-generator-registry.ts:
```diff
@@ -0,0 +1,70 @@
+    return this.kubeGenerators.get(kubeGeneratorId);
+  }
+
+  getKubeGeneratorsInfos(selector: KubernetesGeneratorSelector | undefined = undefined): KubeGeneratorsInfo[] {
```
```suggestion
  getKubeGeneratorsInfos(selector?: KubernetesGeneratorSelector): KubeGeneratorsInfo[] {
```

benoitf commented on packages/preload/src/index.ts:
```diff
@@ -245,9 +246,12 @@ function initExposure(): void {
-  });
+  contextBridge.exposeInMainWorld(
+    'generatePodmanKube',
+    async (engine: string, names: string[], kubeGeneratorId: string | undefined = undefined): Promise<string> => {
```
```suggestion
    async (engine: string, names: string[], kubeGeneratorId?: string): Promise<string> => {
```

benoitf commented on packages/preload/src/index.ts:
```diff
@@ -923,6 +927,13 @@ function initExposure(): void {
 
+  contextBridge.exposeInMainWorld(
+    'getKubeGeneratorsInfos',
+    async (selector: KubernetesGeneratorSelector | undefined = undefined): Promise<string[]> => {
```
```suggestion
    async (selector?: KubernetesGeneratorSelector): Promise<string[]> => {
```

benoitf commented on packages/extension-api/src/extension-api.d.ts:
```diff
@@ -1339,6 +1346,31 @@ declare module '@podman-desktop/api' {
+    id: string;
+    name: string;
+    types: KubernetesGeneratorSelector;
+    generate(engineId: string, ids: string[]): GenerateKubeResult;
```
based on your previous feedback, we need to change string[] to a set of resources like

```ts
{containers: [], pods: [], compose: []}
```

axel7083 replied:
What type for containers pods and compose ? If we use `ContainerInfoUI` we need to add them to the `extension-api.d.ts` ?

benoitf replied:
just IDs so string

axel7083 replied:
composeInfoUI does not have a unique identifier

benoitf replied:
it looks like we submit ids of containers when we're using the generate for compose usecase

benoitf replied:
but we could specify 'com.docker.compose.project' label value ?

axel7083 replied:
> but we could specify 'com.docker.compose.project' label value ?

Humm, I think if we want to use that, we should not be able to send an array of it, I do think the generate function should support an array of compose project.

axel7083 replied:
I will use the containers id for now, as it is in the current implementation

benoitf commented on packages/extension-api/src/extension-api.d.ts:
```diff
@@ -1339,6 +1346,31 @@ declare module '@podman-desktop/api' {
+   * KubernetesGeneratorType.
+   */
+  export interface KubernetesGeneratorProvider {
+    /**
+     * A unique identifier for the provider.
+     */
+    id: string;
```
the unique identifier is not provided by the provider (it's generated/handled by Podman desktop)

```suggestion

```

benoitf commented on packages/main/src/plugin/index.ts:
```diff
@@ -825,8 +829,18 @@ export class PluginSystem {
+      async (
+        _listener,
+        engine: string,
+        names: string[],
```
should ask the resources like {pods, containers and compose}

benoitf commented on packages/main/src/plugin/kube-generator-registry.ts:
```diff
@@ -0,0 +1,70 @@
+  types: KubernetesGeneratorSelector;
+}
+
+export type KubeGeneratorsInfo = Omit<KubernetesGeneratorProvider, 'generate'>;
```
I think it's risky to omit fields

Also the info object needs to be in a separate file as it's imported from the client side

It ensures client will not fetch unwated dependencies if it imports this file

benoitf commented on packages/preload/src/index.ts:
```diff
@@ -245,9 +246,12 @@ function initExposure(): void {
-  });
+  contextBridge.exposeInMainWorld(
+    'generatePodmanKube',
+    async (engine: string, names: string[], kubeGeneratorId?: string): Promise<string> => {
+      return ipcInvoke('container-provider-registry:generatePodmanKube', engine, names, kubeGeneratorId);
+    },
```
should we add for now a new method that return GenerateKubeResult instead of string (and mark this method as deprecated)

then update rendering part to the new API and remove this method after ?

benoitf replied:
I've a question: should we not keep the method as before ? (so no kubeGeneratorId for now) until the renderer part switch to the new method

axel7083 commented on packages/main/src/plugin/kube-generator-registry.ts:
```diff
@@ -0,0 +1,70 @@
+        (Array.isArray(selector) ? selector : [selector]).every(type => generator.types.includes(type));
+
+      if (isMatchingGenerator) {
+        // eslint-disable-next-line @typescript-eslint/no-unused-vars
```
No, the `_` is not ignored, and this made me crazy !

axel7083 commented:
Some changes (again)


## New windows function

Introducing `windows.generateKube(kubeGeneratorId: string, kubernetesGeneratorArgument: KubernetesGeneratorArgument): Promise<GenerateKubeResult>`

This will be the replacement of the existing `generatePodmanKube` in the future.

## New type `KubernetesGeneratorArgument`

The `generate` function of the provider, take a KubernetesGeneratorArgument  object as argument, since we need to able to Omit field
> if we generate for compose, when calling windows.generateKube we do not want to have mandatory arguments that we have no use for.

```typescript
export type KubernetesGeneratorArgument = {
  containers?: string[];
  pods?: string[];
  compose?: string[];
}
```

## ProviderId changes

The providerIds is now defined by the registry, and has been removed from the  `KubernetesGeneratorProvider` interface. 
It is included in the `KubernetesGeneratorInfo` interface.

> The unit tests have been adapted to ensure they are unique

## Usage example

For example let's take the `ComposeDetilsKube.svelte` we can have

```typescript
<script lang="ts">
import { onMount } from 'svelte';
import MonacoEditor from '../editor/MonacoEditor.svelte';
import type { ComposeInfoUI } from './ComposeInfoUI';
import type {KubernetesGeneratorInfo} from "../../../../main/src/plugin/api/KubernetesGeneratorInfo";

export let compose: ComposeInfoUI;

let kubeDetails: string;

onMount(async () => {
  const providersIds: KubernetesGeneratorInfo[] = await window.getKubeGeneratorsInfos('Compose');

  const kubeResult = await window.generateKube(providersIds[0].id, {
    compose: compose.containers.map(container => container.id)
  });
  kubeDetails = kubeResult.yaml;
});
</script>

{#if kubeDetails}
  <MonacoEditor content="{kubeDetails}" language="yaml" />
{/if}
```

And in our extension.ts we got

```typescript
import * as podmanDesktopAPI from '@podman-desktop/api';

export async function activate(extensionContext: podmanDesktopAPI.ExtensionContext): Promise<void> {
  extensionContext.subscriptions.push(
      podmanDesktopAPI.kubernetes.registerKubernetesGenerator({
          types: ['Compose'],
          name: 'Kompose',
          id: 'kompose',
          generate: (
              { compose }: podmanDesktopAPI.KubernetesGeneratorArgument
          ) => {

              return {
                  yaml: `hello from extension received ${compose}`
              };
          }
      }),
  );
}
```

benoitf commented on packages/preload/src/index.ts:
```diff
@@ -245,9 +252,22 @@ function initExposure(): void {
+  contextBridge.exposeInMainWorld(
+    'generateKube',
+    async (
+      kubeGeneratorId: string,
+      kubernetesGeneratorArgument: KubernetesGeneratorArgument,
```
shouldn't we have kubeGeneratorId as last parameter and being optional ?
```suggestion
      kubernetesGeneratorArgument: KubernetesGeneratorArgument,
      kubeGeneratorId?: string,
```

axel7083 replied:
Depends, do we consider Podman generator to be the default one ? and being called when kubeGeneratorId is undefined ?

benoitf replied:
I think that there is room for a default one, maybe could be configured later also but yes generally there should always be a default (if matching the expectation and if available)

axel7083 replied:
How do we define the default, can an extension ask to be the default one ?
In the current implementation, simply called manually the podman `containerProviderRegistry.generatePodmanKube` ? We lost the information about the engine, so we should add it to the kubernetesGeneratorArgument ?

benoitf replied:
so yes I think that if we introduce a property to name the default, we would have the name of the default generator persisted.

then, an extension can read that property as well.

yes it looks like engineId is important so it looks it requires another changes, when you specify the list of pods, compose, and containers, you need to specify both engineId and then the list of containers for this engineId, etc.

```ts
[
   {
      "engineId":"abcd",
      "containers":[
         "1",
         "2",
         "3"
      ],
      "pods":[
         "1"
      ],
      "compose":[
         "2"
      ]
   },
   {
      "engineId":"fgh",
      "pods":[
         "1"
      ]
   }
]
```

benoitf commented:
FYI there is a conflict

axel7083 commented:
I will not have much time to work on this project in the following weeks.

Therefore I included in this PR to register the PodmanKube as the default KubeGenerator.

benoitf commented:
@axel7083 the argument parameter needs to be an array of argument like


```
[
   {
      "engineId":"abcd",
      "containers":[
         "1",
         "2",
         "3"
      ],
      "pods":[
         "1"
      ],
      "compose":[
         "2"
      ]
   },
   {
      "engineId":"fgh",
      "pods":[
         "1"
      ]
   }
]
```

and using varargs we can then only provide one item.

axel7083 commented:
> @axel7083 the argument parameter needs to be an array of argument like
> 
> ```
> [
>    {
>       "engineId":"abcd",
>       "containers":[
>          "1",
>          "2",
>          "3"
>       ],
>       "pods":[
>          "1"
>       ],
>       "compose":[
>          "2"
>       ]
>    },
>    {
>       "engineId":"fgh",
>       "pods":[
>          "1"
>       ]
>    }
> ]
> ```
> 
> and using varargs we can then only provide one item.

I do not understand why, would we want to take multiple `KubernetesGeneratorArgument` to the `windows.generateKube`, the existing `windows.generatePodmanKube` only takes `{engine: string, names: string[]}` as argument.

benoitf commented:
you ask from client side for a set of containers, pods and compose. 
These elements can be on different elements
So client ask in one command everything

Then on main side, it checks if IDs are ok for the given IDs,etc

And then call providers for each engine I'd

Then it respond the Yall

If we don't use multiple engines it means client will prevent users to select containers on different engines

axel7083 commented:
> you ask from client side for a set of containers, pods and compose. These elements can be on different elements So client ask in one command everything
> 
> Then on main side, it checks if IDs are ok for the given IDs,etc
> 
> And then call providers for each engine I'd
> 
> Then it respond the Yall
> 
> If we don't use multiple engines it means client will prevent users to select containers on different engines

Okey yeah that make sense, I have not thought to some scenario.

Some issues still related with the arguments orders. Since the kubeGeneratorId is an optional arguments (undefined => using default). You told me to place it at the end of the functions arguments (make sense).

```js
contextBridge.exposeInMainWorld(
    'generateKube',
    async (
      kubernetesGeneratorArgument: KubernetesGeneratorArgument,
      kubeGeneratorId?: string,
    ): Promise<GenerateKubeResult> => {
      return ipcInvoke('kubernetes-generator-registry:generateKube', kubernetesGeneratorArgument, kubeGeneratorId);
    },
  );
```

If we use varargs, where do we place the kubeGeneratorId ? We would have to make it mandatory. 
Another solution would be to change kubernetesGeneratorArgument to `kubernetesGeneratorArguments: KubernetesGeneratorArgument[]`

benoitf commented:
@axel7083 ok let's not use varargs then
